ml-conversational-analytic-tool/githubDataExtraction.py

```diff
@@ -1,179 +1,240 @@
-            except RateLimitExceededException as e:  # getPullFeatures accesses the Github API so provisino for rate limit
-                print("Token rate limit exceeded. Waiting for 1 hour", e)
-                time.sleep(60 * 60)
-                pull_data.append(self.getPullFeatures(pull))
-        return pull_data
-
-    def listOfComments(self, comments):
+        has_next_page = True
+        data = []
+
+        while has_next_page:
+            # wait two seconds to avoid the secondary rate limits
+            time.sleep(2)
+            print('retrieving data...')
+
+            try:
+                response = requests.post(url='https://api.github.com/graphql', json=query, headers=self.headers)
+                response_json = response.json()
+                response_reason = response.reason
+                status_code = response.status_code
+
+                if status_code == 200:
+                    # GitHub graphql has a bug that returns status 200 for exceeding the rate limit
+                    # When a rate limit has exceeded, retries after waiting until rate limit has reset
+                    # The wait time is calculated using returned header X-RateLimit-Reset
+                    rate_limit_remaining = response.headers.get('X-RateLimit-Remaining')
+                    if rate_limit_remaining == "0":
+                        rate_limit_reset = response.headers.get('X-RateLimit-Reset')
+                        rate_limit_reset_time = datetime.datetime.fromtimestamp(int(rate_limit_reset))
+                        current_time = datetime.datetime.now()
+                        wait_time_seconds = (rate_limit_reset_time - current_time).total_seconds()
+
+                        print('rate limit exceeded, retrying after {} seconds'.format(wait_time_seconds))
+                        time.sleep(wait_time_seconds)
+                    else:
+                        if query['query'] == self.query_repos:
+                            org_repos = response_json['data']['organization']['repositories']
+                            page_info = org_repos['pageInfo']
+                            nodes = org_repos['nodes']
+                        else:
+                            repo_pull_requests = response_json['data']['repository']['pullRequests']
+                            page_info = repo_pull_requests['pageInfo']
+                            nodes = repo_pull_requests['nodes']
+
+                        end_cursor = page_info['endCursor']
+                        has_next_page = page_info['hasNextPage']
+
+                        variables = query['variables']
+                        variables['cursor'] = end_cursor
+                        query = {'query': query['query'], 'variables': variables}
+
+                        data += nodes
+                elif status_code == 502 or status_code == 503 or status_code == 403 or status_code == 413 or status_code == 429:
```

Comment:
nit: what about rewriting this line to `elis status_code in (...)` - to be shorter?

Reply:
Done